psh/hm/hm.c

```diff
@@ -121,20 +126,149 @@ void psh_hminfo(void)
 }
 
 
+static int psh_hm_cmppid(const void *t1, const void *t2)
+{
+	return (int)((threadinfo_t *)t1)->pid - (int)((threadinfo_t *)t2)->pid;
+}
+
+
+static void psh_hm_freeProcesses(rbtree_t *ptree)
+{
+	rbnode_t *n;
+	proc_t *p;
+
+	for (n = ptree->root; n != NULL; n = ptree->root) {
+		p = lib_treeof(proc_t, node, n);
+		lib_rbRemove(ptree, n);
+		free(p);
+	}
+}
+
+
+static int psh_hm_getProcesses(rbtree_t *ptree)
+{
+	int tcnt, i, j, n = 32;
+	threadinfo_t *info, *rinfo;
+	proc_t *p;
+
+	if ((info = malloc(n * sizeof(threadinfo_t))) == NULL) {
+		fprintf(stderr, "hm: Out of memory\n");
+		return -ENOMEM;
+	}
+
+	while ((tcnt = threadsinfo(n, info)) >= n) {
+		n *= 2;
+		if ((rinfo = realloc(info, n * sizeof(threadinfo_t))) == NULL) {
+			fprintf(stderr, "hm: Out of memory\n");
+			free(info);
+			return -ENOMEM;
+		}
+		info = rinfo;
+	}
+
+	qsort(info, tcnt, sizeof(threadinfo_t), psh_hm_cmppid);
+
+	i = 0;
+
+	while (i < tcnt) {
```

Comment:
Can't we just use threadinfo_t tables? This way we need to perform n iterations to copy table to the tree and also perform n memory allocations. Comparing two sorted tables should take in most cases O(1) (we have a hint after finding last one - next one is not bigger, than the previous one)